pkg/pillar/hypervisor/hypervisor.go

```diff
@@ -38,17 +38,17 @@ type Hypervisor interface {
 
 type hypervisorDesc struct {
 	constructor       func() Hypervisor
-	dom0handle        string
+	enabled           bool
```

Comment:
You're evaluating the existence when the Zedbox process starts and containerd etc might not be running that early.

enabled func() bool
would keep the time of the check the same.

Reply:
My comment still holds - why change the assumptions and behavior as part of this refactoring?

Reply:
I did not understand your comment. What behavior did I change ? May be we can chat offline.

Reply:
If I understood Erik correctly, he is saying that the current behaviour just lists the file to check in the start of the program, but it does not actually check it until you call `GetAvailableHypervisors()`, which might be much later. With this change, you check it on initial startup. If one of the device files or sockets is not there yet, you will miss it.

Is this system containerd? Why would system containerd be looked at as part of the pillar hypervisor list? pillar doesn't touch it; all content and workloads are via user containerd. If it actually _is_ looking at system containerd, that is a bug.

On the other hand, I really like what you did about giving each entry its own logic. They vary so much between them, it makes more sense to keep the logic for checking each as part of it, and cleans it up a lot.

What Erik is recommending is, I believe, the following:

```go
type hypervisorDesc struct {
	constructor       func() Hypervisor
	enabled           func() bool
	hvTypeFileContent string
}

var knownHypervisors = map[string]hypervisorDesc{
	XenHypervisorName:        {constructor: newXen, enabled: func () bool {return fileutils.FileExists(nil, "/proc/xen")}, hvTypeFileContent: "xen"},
	KVMHypervisorName:        {constructor: newKvm, enabled: func() bool{fileutils.FileExists(nil, "/dev/kvm") && !base.IsHVTypeKube()}, hvTypeFileContent: "kvm"},
	KubevirtHypervisorName:   {constructor: newKubevirt, enabled: func() bool{ fileutils.FileExists(nil, "/dev/kvm") && base.IsHVTypeKube()}, hvTypeFileContent: "kubevirt"},
	ACRNHypervisorName:       {constructor: newAcrn, enabled: func() bool {fileutils.FileExists(nil, "/dev/acrn")}, hvTypeFileContent: "acrn"},
	ContainerdHypervisorName: {constructor: newContainerd, enabled: func() bool{ fileutils.FileExists(nil, "/run/containerd/containerd.sock")|},
	NullHypervisorName:       {constructor: newNull, enabled: func() bool{fileutils.DirExists(nil, "/")}},
}

func GetAvailableHypervisors() (all []string, enabled []string) {
	all = hypervisorPriority
	for _, v := range all {
		if knownHypervisors[v].enabled() {
			enabled = append(enabled, v)
		}
	}

}
```

Reply:
Correct @deitch

Reply:
Thanks @deitch I get it now